buildSrc/src/main/groovy/org/elasticsearch/gradle/VersionCollection.groovy

```diff
@@ -25,159 +25,293 @@ import java.util.regex.Matcher
+            compatSnapshots.add(stagedMinorSnapshot)
+        }
+        if (allCompatVersions.contains(nextBugfixSnapshot)) {
+            compatSnapshots.add(nextBugfixSnapshot)
         }
-        return snapshotIndex == -1 ? null : versions[snapshotIndex]
+        if (allCompatVersions.contains(maintenanceBugfixSnapshot)) {
+            compatSnapshots.add(maintenanceBugfixSnapshot)
+        }
+        // There was no wire compat for the 2.x line
+        compatSnapshots.removeAll {it.major == 2}
+
+        return compatSnapshots;
     }
 
-    private List<Version> versionsOnOrAfterExceptCurrent(Version minVersion) {
-        final String minVersionString = minVersion.toString()
-        return Collections.unmodifiableList(versions.findAll {
-            it.onOrAfter(minVersionString) && it != currentVersion
-        })
+    /**
+     * Uses basic logic about our releases to determine if this version has been previously released
+     */
+    private boolean isReleased(Version version) {
+        return version.revision > 0 || (version.revision > 1 && currentVersion.equals(Version.fromString("5.1.2")))
     }
 
     /**
-     * @return All earlier versions that should be tested for index BWC with the current version.
+     * Validates that the count of non suffixed (alpha/beta/rc) versions in a given major to major+1 is greater than 1.
+     * This means that there is more than just a major.0.0 or major.0.0-alpha in a branch to signify it has been prevously released.
      */
-    List<Version> getVersionsIndexCompatibleWithCurrent() {
-        final Version firstVersionOfCurrentMajor = versions.find { it.major >= currentVersion.major - 1 }
-        return versionsOnOrAfterExceptCurrent(firstVersionOfCurrentMajor)
+    private boolean isMajorReleased(Version version, TreeSet<Version> items) {
+        return items
+            .tailSet(Version.fromString("${version.major}.0.0"))
+            .headSet(Version.fromString("${version.major + 1}.0.0"))
+            .count { it.suffix.isEmpty() }  // count only non suffix'd versions as actual versions that may be released
+            .intValue() > 1
     }
 
-    private Version getMinimumWireCompatibilityVersion() {
-        final int firstIndexOfThisMajor = versions.findIndexOf { it.major == currentVersion.major }
-        if (firstIndexOfThisMajor == 0) {
-            return versions[0]
+    /**
+     * Gets the largest version previous major version based on the nextMajorVersion passed in.
+     * If you have a list [5.0.2, 5.1.2, 6.0.1, 6.1.1] and pass in 6 for the nextMajorVersion, it will return you 5.1.2
+     */
+    private Version getHighestPreviousMinor(Integer nextMajorVersion) {
+        return versionSet.headSet(Version.fromString("${nextMajorVersion}.0.0")).last()
+    }
+
+    /**
+     * Helper function for turning a version into a snapshot version, removing and readding it to the tree
+     */
+    private Version removeAndReaddAsSnapshot(Version version) {
```

Comment:
Typo: too many `d`s.

Reply:
thats actually Re-Add, which after re-reading, is a terrible name.

Reply:
Why not just `replaceAsSnapshot`?